tensorpipe/channel/cuda/channel.cc

```diff
@@ -0,0 +1,406 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/channel/cuda/channel.h>
+
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <algorithm>
+#include <limits>
+#include <list>
+
+#include <cuda_runtime.h>
+
+#include <tensorpipe/channel/error.h>
+#include <tensorpipe/channel/helpers.h>
+#include <tensorpipe/common/callback.h>
+#include <tensorpipe/common/defs.h>
+#include <tensorpipe/common/error.h>
+#include <tensorpipe/common/error_macros.h>
+#include <tensorpipe/common/optional.h>
+#include <tensorpipe/common/queue.h>
+#include <tensorpipe/common/system.h>
+#include <tensorpipe/proto/channel/cuda.pb.h>
+
+namespace tensorpipe {
+namespace channel {
+namespace cuda {
+
+class Channel::Impl : public std::enable_shared_from_this<Channel::Impl> {
+ public:
+  Impl(
+      std::shared_ptr<Context::PrivateIface>,
+      std::shared_ptr<transport::Connection>,
+      std::string,
+      cudaStream_t = 0);
```

Comment:
I don't think CUDA guarantees that `cudaStream_t` is a pointer type. Moreover, the docs states in various places (such as [cudaMemcpyAsync](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79)) that a default stream is represented by the value `0`.

Reply:
In the [CUDA docs page on types](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html) I'm seeing:
```
typedef CUstream_st *  cudaStream_t
```
Doesn't that mean it's always a pointer type?

Maybe the CUDA docs use 0 because it's a C APIs and thus it cannot use nullptr (and doesn't want to use NULL because it's somewhat undefined)?

Reply:
Fair point. In any case, we should probably use `cudaStreamDefault` (that I learned about today). Ironically, it is defined to `0`.

Reply:
That's perfect! It's great if we can use a meaningful constant defined by CUDA rather than hardcoding its value ourselves.